src/mlpack/methods/ann/layer/transposed_convolution.hpp

```diff
@@ -246,20 +243,81 @@ class TransposedConvolution
    * @param input The input to be padded.
    * @param wPad Padding width of the input.
    * @param hPad Padding height of the input.
+   * @param wExtra The number of extra zeros to the right.
+   * @param hExtra The number of extra zeros to the bottom.
    * @param output The padded output data.
    */
   template<typename eT>
   void Pad(const arma::Cube<eT>& input,
            size_t wPad,
            size_t hPad,
+           size_t wExtra,
+           size_t hExtra,
            arma::Cube<eT>& output)
   {
-    output = arma::zeros(input.n_rows + wPad * 2,
-        input.n_cols + hPad * 2, input.n_slices);
+    output = arma::zeros(input.n_rows + wPad * 2 + wExtra,
+        input.n_cols + hPad * 2 + hExtra, input.n_slices);
 
     for (size_t i = 0; i < input.n_slices; ++i)
     {
-      Pad<eT>(input.slice(i), wPad, hPad, output.slice(i));
+      Pad<eT>(input.slice(i), wPad, hPad, wExtra, hExtra, output.slice(i));
+    }
+  }
+
+  /*
+   * Insert zeros between the units of the given input data.
+   * Note: This function should be used before the Pad() function.
+   *
+   * @param input The input to be padded.
+   * @param dW Stride of filter application in the x direction.
+   * @param dH Stride of filter application in the y direction.
+   * @param output The padded output data.
+   */
+  template<typename eT>
+  void InsertZeros(const arma::Mat<eT>& input,
+                   size_t dW,
+                   size_t dH,
+                   arma::Mat<eT>& output)
+  {
+    if (output.n_rows != input.n_rows * dW - dW + 1 ||
+        output.n_cols != input.n_cols * dH - dH + 1)
+    {
+      output = arma::zeros(input.n_rows * dW - dW + 1,
+          input.n_cols * dH - dH + 1);
+    }
+
+    for (size_t i = 0; i < output.n_rows; i += dH)
+    {
+      for (size_t j = 0; j < output.n_cols; j += dW)
+      {
+        // TODO: Use [] instead of () for speedup after this is completely
+        // debugged and approved.
+        output(i, j) = input(i / dH, j / dW);
```

Comment:
Not sure using ``[]`` will bring us any speedup here, but maybe I missed something.

Reply:
Isn't `[]` an optimized version of `()` without bounds check? It's mentioned in the Armadillo docs.

Reply:
Right, this should only matter in debug mode:

ARMA_NO_DEBUG |   | Disable all run-time checks, such as bounds checking. This will result in faster code, but you first need to make sure that your code runs correctly! We strongly recommend to have the run-time checks enabled during development, as this greatly aids in finding mistakes in your code, and hence speeds up development. We recommend that run-time checks be disabled only for the shipped version of your program (ie. final release build).
-- | -- | --

Reply:
Oh, I didn't know that. I guess I'll leave it with `()` then, might actually be useful in debug mode if something goes wrong.